src/StreamJsonRpc/SystemTextJsonFormatter.cs

```diff
@@ -0,0 +1,1414 @@
+                }
+                else
+                {
+                    typeInfo = this.fallbackResolver.GetTypeInfo(type, options);
+                }
+
+                this.typeInfoCache.TryAdd(type, typeInfo);
+            }
+
+            return typeInfo;
+        }
+
+        private static void PopulateMembersInfos(Type type, JsonTypeInfo jsonTypeInfo, DataContractAttribute? dataContractAttribute)
+        {
+            BindingFlags bindingFlags = BindingFlags.Public | BindingFlags.Instance;
+
+            // When the type is decorated with DataContractAttribute, we can consider non-public members.
+            if (dataContractAttribute is not null)
+            {
+                bindingFlags |= BindingFlags.NonPublic;
+            }
+
+            foreach (PropertyInfo propertyInfo in type.GetProperties(bindingFlags))
+            {
+                if (TryCreateJsonPropertyInfo(propertyInfo, propertyInfo.PropertyType, out JsonPropertyInfo? jsonPropertyInfo))
+                {
+                    if (propertyInfo.CanRead)
+                    {
+                        jsonPropertyInfo.Get = propertyInfo.GetValue;
+                    }
+
+                    if (propertyInfo.CanWrite)
+                    {
+                        jsonPropertyInfo.Set = propertyInfo.SetValue;
+                    }
+                }
+            }
+
+            foreach (FieldInfo fieldInfo in type.GetFields(bindingFlags))
+            {
+                if (TryCreateJsonPropertyInfo(fieldInfo, fieldInfo.FieldType, out JsonPropertyInfo? jsonPropertyInfo))
+                {
+                    jsonPropertyInfo.Get = fieldInfo.GetValue;
+                    if (!fieldInfo.IsInitOnly)
+                    {
+                        jsonPropertyInfo.Set = fieldInfo.SetValue;
+                    }
+                }
+            }
+
+            bool TryCreateJsonPropertyInfo(MemberInfo memberInfo, Type propertyType, [NotNullWhen(true)] out JsonPropertyInfo? jsonPropertyInfo)
+            {
+                DataMemberAttribute? dataMemberAttribute = memberInfo.GetCustomAttribute<DataMemberAttribute>();
+                if (dataContractAttribute is null || dataMemberAttribute is not null)
+                {
+                    jsonPropertyInfo = jsonTypeInfo.CreateJsonPropertyInfo(propertyType, dataMemberAttribute?.Name ?? memberInfo.Name);
+                    if (dataMemberAttribute is not null)
+                    {
+                        jsonPropertyInfo.Order = dataMemberAttribute.Order;
+                        jsonPropertyInfo.IsRequired = dataMemberAttribute.IsRequired;
+                        if (!dataMemberAttribute.EmitDefaultValue)
+                        {
+                            object? defaultValue = propertyType.IsValueType ? FormatterServices.GetUninitializedObject(propertyType) : null;
+                            jsonPropertyInfo.ShouldSerialize = (_, value) => !object.Equals(defaultValue, value);
```

Comment:
There may not be much we can do as long as we rely on the IJsonTypeInfoResolver for this, but this will result in boxing upon serialization of all value types property with `EmitDefaultValue == false` which is very common for booleans.
We can consider a future optimization where we address this replacing the `IJsonTypeInfoResolver` with a `JsonConverterFactory`

Reply:
Ooh, that's a good point. I'll brainstorm on this a bit more.

Reply:
The same applies to `typeInfo.CreateObject = () => FormatterServices.GetUninitializedObject(type);` above BTW

Reply:
Nothing we can do about `CreateObject`, since it is typed as `Func<object>`

Reply:
In fact, `ShouldSerialize` is also in the same boat: the func signature forces the value to have *already* been boxed. We're not boxing it, System.Text.Json is. We can only hope they have a similar optimization to what newtonsoft recently picked up: recognizing common, small values for value-types and reuse the same boxes.